File: python/p0020.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []  # Stack to keep track of opening parentheses

        for char in s:
            if char in "({[":
                stack.append(char)  # Push opening parentheses onto the stack
            else:
                if not stack:
                    return False  # No matching opening parenthesis

                last = stack.pop()  # Pop the last opening parenthesis

                # Check for matching parentheses
                if (char == ")" and last != "(") or \
                (char == "]" and last != "[") or \
                (char == "}" and last != "{"):
                    return False

        return not stack  # All parentheses matched

File: javascript/p0020.js
/**
 * @param {string} s
 * @return {boolean}
 */
export var isValid = function (s) {
  let stack = []; // Keep track of opening parentheses

  for (let i = 0; i < s.length; i++) {
    let char = s[i];

    if (char == "(" || char == "[" || char == "{") {
      stack.push(char); // Push opening parentheses onto the stack
    } else {
      if (stack.length === 0) return false; // No matching opening parenthesis

      let last = stack.pop(); // Pop the last opening parenthesis

      // Check for matching parentheses
      if (
        (char == ")" && last != "(") ||
        (char == "]" && last != "[") ||
        (char == "}" && last != "{")
      ) {
        return false;
      }
    }
  }

  return stack.length === 0; // All parentheses matched
};
